deadlines/src/main/java/com/palantir/deadlines/DeadlineExpiredReasons.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.deadlines;
+
+import java.util.Optional;
+
+public final class DeadlineExpiredReasons {
+
+    private DeadlineExpiredReasons() {}
+
+    public static <T> void encodeToResponse(
+            DeadlineExpiredException exception, T response, ResponseEncodingAdapter<T> adapter) {
+        if (exception instanceof DeadlineExpiredException.External) {
+            adapter.setStatus(response, 400);
+            adapter.setHeader(response, DeadlinesHttpHeaders.DEADLINE_EXPIRED_REASON, "external");
+        } else if (exception instanceof DeadlineExpiredException.Internal) {
+            adapter.setStatus(response, 500);
+            adapter.setHeader(response, DeadlinesHttpHeaders.DEADLINE_EXPIRED_REASON, "internal");
+        } else {
+            adapter.setStatus(response, 500);
+        }
+    }
+
+    public static <T> Optional<DeadlineExpiredException> parseFromResponse(
+            T response, ResponseDecodingAdapter<T> adapter) {
+        Optional<String> reason = adapter.getFirstHeader(response, DeadlinesHttpHeaders.DEADLINE_EXPIRED_REASON);
+        return reason.flatMap(s -> switch (s) {
+            case "external" -> Optional.of(DeadlineExpiredException.external());
+            case "internal" -> Optional.of(DeadlineExpiredException.internal());
+            default -> Optional.empty();
+        });
+    }
+
+    public interface ResponseEncodingAdapter<RESPONSE> {
+        void setHeader(RESPONSE response, String headerName, String headerValue);
+
+        void setStatus(RESPONSE response, int status);
```

Comment:
perhaps we should invert this and use the visitor style a la QosException

Reply:
to expand on that: reason being that in e.g. https://github.com/palantir/conjure-java/pull/2454, we're leveraging the private `ConjureExceptions#writeResponse` which accepts an http status as a parameter, but also contains useful logic that we probably want to re-use involved in writing the response body.

instead of including a `setStatus()` method on the encoder, we could expose a basic method here:
```
public int httpStatusFor(DeadlineExpiredException exception) {
    if (exception instanceof DeadlineExpiredException.External) {
        return 400;
    } else if (exception instanceof DeadlineExpiredException.Internal) {
        return 500;
    }
    return 500;
}
```

thoughts?

Reply:
I think either way could work. I don't think we necessarily need to send a response at all (similarly to QosException), although I suppose an `errorInstanceId` might be nice to quickly find the failures propagation path back to clients.

Reply:
updated to use the visitor thing. i don't love it... but i guess it works